### PR TITLE
Support for persistence mode in L2 networks

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR aims at introducing persistence mode in L2 networks and enhancing the behavior in Isolated networks

Doc PR https://github.com/apache/cloudstack-documentation/pull/183

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/command/user/network/CreateNetworkCmd.java [hunk 1]
```diff
@@ -48,7 +49,7 @@
 
 @APICommand(name = "createNetwork", description = "Creates a network", responseObject = NetworkResponse.class, responseView = ResponseView.Restricted, entityType = {Network.class},
         requestHasSensitiveInfo = false, responseHasSensitiveInfo = false)
-public class CreateNetworkCmd extends BaseCmd implements UserCmd {
+public class CreateNetworkCmd extends BaseAsyncCmd implements UserCmd {
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/command/user/network/CreateNetworkCmd.java:52]: @Pearl1594 can you start a discussion thread on dev@/users@ we want to see if it causes any backward compatibility issues for users

Comment (nvazquez) [reply api/src/main/java/org/apache/cloudstack/api/command/user/network/CreateNetworkCmd.java:52]: +1

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 2]
```diff
@@ -1187,6 +1200,68 @@ boolean isNetworkImplemented(final NetworkVO network) {
...
+            if (isNtwConfiguredInCluster(host, clusterToHostsMap)) {
+                continue;
+            }
+            final SetupPersistentNetworkAnswer answer = (SetupPersistentNetworkAnswer)_agentMgr.send(host.getId(), cmd);
+
+            if (answer == null) {
+                s_logger.warn("Unable to get an answer to the SetupPersistentNetworkCommand from agent:" + host.getId());
+                clusterToHostsMap.get(host.getClusterId()).remove(host.getId());
+                continue;
+                // throw new CloudRuntimeException("Unable to get an answer to the SetupPersistentNetworkCommand from agent: " + host.getId());
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1250]: Please remove

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 3]
```diff
@@ -1187,6 +1200,68 @@ boolean isNetworkImplemented(final NetworkVO network) {
...
+            }
+
+            if (!answer.getResult()) {
+                s_logger.warn("Unable to setup agent " + host.getId() + " due to " + answer.getDetails() );
+//                final String msg = "Incorrect Network setup on agent, Reinitialize agent after network names are setup, details : " + answer.getDetails();
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1255]: Same here

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 4]
```diff
@@ -1187,6 +1200,68 @@ boolean isNetworkImplemented(final NetworkVO network) {
...
+        to.setBroadcastType(networkVO.getBroadcastDomainType());
+        to.setType(networkVO.getTrafficType());
+        to.setBroadcastUri(networkVO.getBroadcastUri());
+        to.setIsolationuri(networkVO.getBroadcastUri());
+        to.setNetworkRateMbps(_configMgr.getNetworkOfferingNetworkRate(networkOfferingVO.getId(), networkVO.getDataCenterId()));
+        to.setSecurityGroupEnabled(_networkModel.isSecurityGroupSupportedInNetwork(networkVO));
+        return to;
+    }
+
+    private boolean isNtwConfiguredInCluster(HostVO hostVO, Map<Long, List<Long>> clusterToHostsMap) {
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1214]: I think this method needs an explanatory javadoc as well

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 5]
```diff
@@ -1187,6 +1200,68 @@ boolean isNetworkImplemented(final NetworkVO network) {
         return implemented;
     }
 
+    private NicTO getNicTO(NetworkVO networkVO, NetworkOfferingVO networkOfferingVO) {
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1203]: Maybe renamed to `createNicTOFromNetworkAndOffering()` or similar?

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 6]
```diff
@@ -1187,6 +1200,68 @@ boolean isNetworkImplemented(final NetworkVO network) {
...
+        hosts.add(hostVO.getId());
+        clusterToHostsMap.put(clusterId, hosts);
+        return false;
+    }
+
+    private void setupPersistentNetwork(NetworkVO network, NetworkOfferingVO offering, Long dcId) throws AgentUnavailableException, OperationTimedoutException {
+        NicTO to = getNicTO(network, offering);
+        List<ClusterVO> clusterVOS = clusterDao.listClustersByDcId(dcId);
+        List<HostVO> hosts = resourceManager.listAllUpAndEnabledHostsInOneZoneByType(Host.Type.Routing, dcId);
+        Collections.reverse(hosts);
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1237]: Why is the list reversed at this point?

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 7]
```diff
@@ -3512,7 +3590,7 @@ private boolean cleanupNetworkResources(final long networkId, final Account call
             }
         } catch (final ResourceUnavailableException ex) {
             success = false;
-            s_logger.warn("Failed to cleanup Network ACLs as a part of network id=" + networkId + " cleanup due to resourceUnavailable ", ex);
+            s_logger.warn("Failed to c*.javaleanup Network ACLs as a part of network id=" + networkId + " cleanup due to resourceUnavailable ", ex);
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:3593]: Small typo

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 8]
```diff
@@ -276,18 +277,33 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
     }
 
+    private String getVlanIdFromBridgeName(String brName) {
+        if (brName != null) {
+            return brName.split("-")[1];
+        }
+        return null;
+    }
+
+    private boolean deleteBridge(Map<String, Boolean> vlanToPersistenceMap, String vlanId) {
```
Comment (nvazquez) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:300]: Maybe renamed to `shouldDeleteBridge()` or similar?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java [hunk 9]
```diff
@@ -276,18 +277,33 @@ Use VIR_DOMAIN_XML_SECURE (value = 1) prior to v1.0.0.
...
+                    vifDriver.unplug(iface, deleteBridge(vlanToPersistenceMap, vlanId));
                 }
             }
         }
 
         return new MigrateAnswer(command, result == null, result, null);
     }
 
+    private String getVlanIdFromBridgeName(String brName) {
+        if (brName != null) {
```
Comment (nvazquez) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtMigrateCommandWrapper.java:294]: What about `StringUtils.isNotBlank()` to also prevent empty strings?

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 10]
```diff
@@ -1374,10 +1374,10 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
         // if the network offering has persistent set to true, implement the network
         if (ntwkOff.isPersistent()) {
             try {
-                if (network.getState() == Network.State.Setup) {
-                    s_logger.debug("Network id=" + network.getId() + " is already provisioned");
-                    return network;
-                }
+//                if (network.getState() == Network.State.Setup) {
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1377]: Please remove

**Review by nvazquez — COMMENTED** [review]
Looks good, leaving some review comments

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matix

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan test matrix

File: api/src/main/java/org/apache/cloudstack/api/command/user/network/CreateNetworkCmd.java [hunk 11]
```diff
@@ -16,13 +16,13 @@
...
 package org.apache.cloudstack.api.command.user.network;
 
+import org.apache.cloudstack.api.BaseCmd;
 import org.apache.log4j.Logger;
 
 import org.apache.cloudstack.acl.RoleType;
 import org.apache.cloudstack.api.APICommand;
 import org.apache.cloudstack.api.ApiConstants;
 import org.apache.cloudstack.api.ApiErrorCode;
-import org.apache.cloudstack.api.BaseCmd;
```
Comment (rhtyd) [inline api/src/main/java/org/apache/cloudstack/api/command/user/network/CreateNetworkCmd.java:25]: Import reshuffle does not look sortable. You can fix your intellij settings or get preset settings (mine are here https://github.com/rhtyd/dotfiles/blob/master/intellij/settings.jar)

File: core/src/main/java/com/cloud/agent/api/CleanupPersistentNetworkResourceCommand.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package com.cloud.agent.api;
+
+import com.cloud.agent.api.to.NicTO;
+
+public class CleanupPersistentNetworkResourceCommand extends Command{
```
Comment (rhtyd) [inline core/src/main/java/com/cloud/agent/api/CleanupPersistentNetworkResourceCommand.java:22]: minor nit - space after `Command {`

File: core/src/main/java/com/cloud/agent/api/MigrateCommand.java [hunk 13]
```diff
@@ -40,6 +40,7 @@
     private boolean executeInSequence = false;
     private List<MigrateDiskInfo> migrateDiskInfoList = new ArrayList<>();
     private Map<String, DpdkTO> dpdkInterfaceMapping = new HashMap<>();
+    Map<String, Boolean> vlanToPersistenceMap;
```
Comment (rhtyd) [inline core/src/main/java/com/cloud/agent/api/MigrateCommand.java:43]: Just a note - if field is not initialised, the consumer of getter must do a NPE check.

File: core/src/main/java/com/cloud/agent/api/SetupPersistentNetworkCommand.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package com.cloud.agent.api;
+
+import com.cloud.agent.api.to.NicTO;
+
+public class SetupPersistentNetworkCommand extends Command{
```
Comment (rhtyd) [inline core/src/main/java/com/cloud/agent/api/SetupPersistentNetworkCommand.java:22]: minor nit - same `Command {`

File: core/src/main/java/com/cloud/agent/api/StopCommand.java [hunk 15]
```diff
@@ -36,6 +36,7 @@
     String controlIp = null;
     boolean forceStop = false;
     private Map<String, DpdkTO> dpdkInterfaceMapping;
+    Map<String, Boolean> vlanToPersistenceMap;
```
Comment (rhtyd) [inline core/src/main/java/com/cloud/agent/api/StopCommand.java:39]: same as before, NPE check for consumer

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 16]
```diff
@@ -1640,7 +1657,11 @@ private void unmanageVMNics(VirtualMachineProfile profile, VMInstanceVO vm) {
 
     protected boolean sendStop(final VirtualMachineGuru guru, final VirtualMachineProfile profile, final boolean force, final boolean checkBeforeCleanup) {
         final VirtualMachine vm = profile.getVirtualMachine();
+        Map<String, Boolean> vlanToPersistenceMap = getVlanToPersistenceMapForVM(vm.getId());
         StopCommand stpCmd = new StopCommand(vm, getExecuteInSequence(vm.getHypervisorType()), checkBeforeCleanup);
+        if (MapUtils.isNotEmpty(vlanToPersistenceMap)) {
+            stpCmd.setVlanToPersistenceMap(vlanToPersistenceMap);
```
Comment (rhtyd) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1663]: minor nit - could be simply made to set the `getVlanToPersistenceMapForVM` output

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 17]
```diff
@@ -1827,6 +1848,63 @@ private void orchestrateStop(final String vmUuid, final boolean cleanUpEvenIfUna
...
+     * @return <VlanId, ShouldKVMBridgeBeDeleted> - basically returns the vlan ID which is used to determine the
+     * bridge name for KVM hypervisor and based on the network and isolation type and persistent setting of the offering
+     * we decide whether the bridge is to be deleted (KVM) if the last VM in that host is destroyed / migrated
+     */
+    private Pair<String, Boolean> getVMNetworkDetails(NetworkVO networkVO, boolean isPersistent) {
+        URI broadcastUri = networkVO.getBroadcastUri();
+        if (broadcastUri != null) {
+            String scheme = broadcastUri.getScheme();
+            String vlanId = Networks.BroadcastDomainType.getValue(broadcastUri);
+            boolean shouldDelete = !((networkVO.getGuestType() == Network.GuestType.L2 || networkVO.getGuestType() == Network.GuestType.Isolated) &&
```
Comment (rhtyd) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1894]: What about vxlan, @Pearl1594 are persistent networks support for VXLAN isolation?

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 18]
```diff
@@ -2864,6 +2962,33 @@ public boolean shutdownNetworkElementsAndResources(final ReservationContext cont
...
+                        CleanupPersistentNetworkResourceAnswer answer = (CleanupPersistentNetworkResourceAnswer) _agentMgr.send(host.getId(), cmd);
+                        if (answer == null) {
+                            s_logger.warn("Unable to get an answer to the CleanupPersistentNetworkResourceCommand from agent:" + host.getId());
+                        }
+
+                        if (!answer.getResult()) {
+                            s_logger.warn("Unable to setup agent " + host.getId() + " due to " + answer.getDetails());
+                        }
+                    } catch (Exception e) {
+                        s_logger.warn("Failed to cleanup network resources");
```
Comment (rhtyd) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:2985]: Can we log on which host it was failed, or one of the warning/log above?

File: plugins/api/discovery/src/main/java/org/apache/cloudstack/discovery/ApiDiscoveryServiceImpl.java [hunk 19]
```diff
@@ -67,7 +68,7 @@ public boolean start() {
         if (s_apiNameDiscoveryResponseMap == null) {
             long startTime = System.nanoTime();
             s_apiNameDiscoveryResponseMap = new HashMap<String, ApiDiscoveryResponse>();
-            Set<Class<?>> cmdClasses = new HashSet<Class<?>>();
+            Set<Class<?>> cmdClasses = new LinkedHashSet<Class<?>>();
```
Comment (rhtyd) [inline plugins/api/discovery/src/main/java/org/apache/cloudstack/discovery/ApiDiscoveryServiceImpl.java:71]: minor minor nit - The other PR/fix has come here.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/BridgeVifDriver.java [hunk 20]
```diff
@@ -376,6 +376,7 @@ private void deleteVnetBr(String brName) {
             command.add("-v", vNetId);
             command.add("-p", pName);
             command.add("-b", brName);
+            command.add("-d", String.valueOf(deleteBr));
```
Comment (rhtyd) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/BridgeVifDriver.java:379]: should `deleteBr` be used to add the -d command?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/BridgeVifDriver.java [hunk 21]
```diff
@@ -436,4 +437,20 @@ public boolean isExistingBridge(String bridgeName) {
...
+        String trafficLabel = nic.getName();
+        String pifName = _pifs.get(trafficLabel);
+        if (pifName == null) {
+            // if not found in bridge map, maybe traffic label refers to pif already?
+            File pif = new File("/sys/class/net/" + trafficLabel);
+            if (pif.isDirectory()) {
+                pifName = trafficLabel;
+            }
+        }
+        String brName = generateVnetBrName(pifName, vlanId);
```
Comment (rhtyd) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/BridgeVifDriver.java:453]: should null check be done here again?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtCleanupPersistentNetworkResourceCommandWrapper.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+import com.cloud.resource.ResourceWrapper;
+
+@ResourceWrapper(handles = CleanupPersistentNetworkResourceCommand.class)
+public class LibvirtCleanupPersistentNetworkResourceCommandWrapper extends CommandWrapper<CleanupPersistentNetworkResourceCommand, Answer, LibvirtComputingResource> {
+    private static final Logger s_logger = Logger.getLogger(LibvirtCleanupPersistentNetworkResourceCommandWrapper.class);
+    @Override
+    public Answer execute(CleanupPersistentNetworkResourceCommand command, LibvirtComputingResource serverResource) {
+        NicTO nic = command.getNicTO();
+        VifDriver driver = serverResource.getVifDriver(nic.getType());
+        if (driver instanceof BridgeVifDriver) {
```
Comment (rhtyd) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtCleanupPersistentNetworkResourceCommandWrapper.java:39]: Is persistent network/L2 not supported for OVS bridge?

**Review by rhtyd — APPROVED** [review]
LGTM overall, needs testing. Left comments.

File: core/src/main/java/com/cloud/agent/api/StopCommand.java (hunk 15 above)
Comment (Pearl1594) [reply core/src/main/java/com/cloud/agent/api/StopCommand.java:39]: There's an explicit check done at the consuming end to ensure this isn't null

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 17 above)
Comment (Pearl1594) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1894]: L2 persistence mode isn't supported for VxLAN isolation type

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/BridgeVifDriver.java (hunk 20 above)
Comment (Pearl1594) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/BridgeVifDriver.java:379]: yes @rhtyd

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtCleanupPersistentNetworkResourceCommandWrapper.java (hunk 22 above)
Comment (Pearl1594) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtCleanupPersistentNetworkResourceCommandWrapper.java:39]: L2 persistence mode hasn't been extended to support it on ovs for KVM

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 23]
```diff
@@ -1827,6 +1848,63 @@ private void orchestrateStop(final String vmUuid, final boolean cleanUpEvenIfUna
         advanceStop(vm, cleanUpEvenIfUnableToStop);
     }
 
+    private void getPersistenceMap(Map<String, Boolean> vlanToPersistenceMap, NetworkVO networkVO) {
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1851]: `getPersistenceMap` here doesn't return anything, _updatePersistenceMap_ might be best fit here

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 24]
```diff
@@ -1827,6 +1848,63 @@ private void orchestrateStop(final String vmUuid, final boolean cleanUpEvenIfUna
...
+            Pair<String, Boolean> data = getVMNetworkDetails(networkVO, offeringVO.isPersistent());
+            Boolean shouldDeleteNwResource = (MapUtils.isNotEmpty(vlanToPersistenceMap) && data != null) ? vlanToPersistenceMap.get(data.first()) : null;
+            if (data != null && (shouldDeleteNwResource == null || shouldDeleteNwResource)) {
+                vlanToPersistenceMap.put(data.first(), data.second());
+            }
+        }
+    }
+
+    private Map<String, Boolean> getVlanToPersistenceMapForVM(long vmId) {
+        List<UserVmJoinVO> userVmJoinVOS = userVmJoinDao.searchByIds(vmId);
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1863]:
```suggestion
        List<UserVmJoinVO> userVmJoinVOs = userVmJoinDao.searchByIds(vmId);
```

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 25]
```diff
@@ -1827,6 +1848,63 @@ private void orchestrateStop(final String vmUuid, final boolean cleanUpEvenIfUna
...
+        Map<String, Boolean> vlanToPersistenceMap = new HashMap<>();
+        for (UserVmJoinVO userVmJoinVO : userVmJoinVOS) {
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1865]: to be clear, move this for block to else case of "_userVmJoinVOS.isEmpty()_"

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 26]
```diff
@@ -1187,6 +1202,86 @@ boolean isNetworkImplemented(final NetworkVO network) {
...
+        if (hosts != null && !hosts.isEmpty()) {
+            return new Pair<>(true, createNicTOFromNetworkAndOffering(networkVO, networkOfferingVO, hostVO));
+        }
+        hosts.add(hostVO.getId());
+        clusterToHostsMap.put(clusterId, hosts);
+        return new Pair<>(false, createNicTOFromNetworkAndOffering(networkVO, networkOfferingVO, hostVO));
+    }
+
+    private void setupPersistentNetwork(NetworkVO network, NetworkOfferingVO offering, Long dcId) throws AgentUnavailableException, OperationTimedoutException {
+        List<ClusterVO> clusterVOS = clusterDao.listClustersByDcId(dcId);
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1242]:
```suggestion
        List<ClusterVO> clusterVOs = clusterDao.listClustersByDcId(dcId);
```

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 27]
```diff
@@ -2864,6 +2962,33 @@ public boolean shutdownNetworkElementsAndResources(final ReservationContext cont
...
+            if (networkMeetsPersistenceCriteria(network, offering, true) &&
+                    _networksDao.getOtherPersistentNetworksCount(network.getId(), network.getBroadcastUri().toString(), offering.isPersistent()) == 0) {
+                List<HostVO> hosts = resourceManager.listAllUpAndEnabledHostsInOneZoneByType(Host.Type.Routing, network.getDataCenterId());
+                for (HostVO host : hosts) {
+                    try {
+                        NicTO to = createNicTOFromNetworkAndOffering(network, offering, host);
+                        CleanupPersistentNetworkResourceCommand cmd = new CleanupPersistentNetworkResourceCommand(to);
+                        CleanupPersistentNetworkResourceAnswer answer = (CleanupPersistentNetworkResourceAnswer) _agentMgr.send(host.getId(), cmd);
+                        if (answer == null) {
+                            s_logger.warn("Unable to get an answer to the CleanupPersistentNetworkResourceCommand from agent:" + host.getId());
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:2978]: potential NPE issue below. `continue` missing here ? please check.

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 28]
```diff
@@ -1187,6 +1202,86 @@ boolean isNetworkImplemented(final NetworkVO network) {
...
+                if (answer == null) {
+                    s_logger.warn("Unable to get an answer to the SetupPersistentNetworkCommand from agent:" + host.getId());
+                    clusterToHostsMap.get(host.getClusterId()).remove(host.getId());
+                    continue;
+                }
+
+                if (!answer.getResult()) {
+                    s_logger.warn("Unable to setup agent " + host.getId() + " due to " + answer.getDetails());
+                    clusterToHostsMap.get(host.getClusterId()).remove(host.getId());
+                    continue;
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:1265]: `continue` may not be required here

File: engine/schema/src/main/java/com/cloud/network/dao/NetworkDaoImpl.java [hunk 29]
```diff
@@ -265,6 +276,8 @@ protected void init() {
         join10.and("vpc", join10.entity().getVpcId(), Op.EQ);
         PrivateNetworkSearch.join("vpcgateways", join10, PrivateNetworkSearch.entity().getId(), join10.entity().getNetworkId(), JoinBuilder.JoinType.INNER);
         PrivateNetworkSearch.done();
+
+
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/network/dao/NetworkDaoImpl.java:280]: white spaces here

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtUnPlugNicCommandWrapper.java [hunk 30]
```diff
@@ -82,4 +87,22 @@ public Answer execute(final UnPlugNicCommand command, final LibvirtComputingReso
             }
         }
     }
+
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtUnPlugNicCommandWrapper.java:90]: could see same methods `getVlanIdFromBridgeName()` and `shouldDeleteBridge()` defined/implemented in the classes - LibvirtMigrateCommandWrapper, LibvirtStopCommandWrapper, LibvirtUnPlugNicCommandWrapper.

can you defined these in a common class (may be some util class) and use them across.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 31]
```diff
@@ -617,6 +621,21 @@ public Answer executeRequest(Command cmd) {
...
+        VmwareHypervisorHost host = getHyperHost(getServiceContext());
+        String hostname = null;
+        VmwareContext context = getServiceContext();
+        HostMO hostMO = new HostMO(context, host.getMor());
+
+        try {
+            prepareNetworkFromNicInfo(hostMO, cmd.getNic(), false, null);
+            hostname =  host.getHyperHostName();
+        } catch (Exception e) {
+            return new SetupPersistentNetworkAnswer(cmd, false, "failed to get response");
```
Comment (sureshanaparti) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:634]: add the reason for exception in the message here

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@Pearl1594 you have conflicts

**Comment by rhtyd** [comment]
@Pearl1594 is this ready for merging?
@blueorangutan package

**Comment by Pearl1594** [comment]
It is ready for merging @rhtyd

**Comment by rhtyd** [comment]
@Pearl1594 thnx, pl kick a round of smoketests once packaging is done.

**Comment by Pearl1594** [comment]
will do, thanks @rhtyd

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Comparing last test run and results, LGTM.

**Pull request merged**